docs/examples/userguide/language_basics/casting_python.py

```diff
@@ -0,0 +1,22 @@
+from cython.cimports.cpython.ref import PyObject
+
+def main():
+
+    python_string = "foo"
+
+    # Note that the variables below are automatically inferred
+    # as the correct pointer type that is assigned to them.
+    # They do not need to be typed explicitly.
```

Comment:
Files are not aligned due to comments

Reply:
I would suggest that you move these comments outside the code.

Reply:
I also wonder if we should just rely on type inference in both versions. And then really move the comment into the preceding text section.

Reply:
> I also wonder if we should just rely on type inference in both versions.

Is it possible to use type inference in .pyx version? I am getting error when removing `cdef + TYPE` part:
```cython
from cpython.ref cimport PyObject

cdef extern from *:
    ctypedef Py_ssize_t Py_intptr_t

python_string = "foo"

ptr = <void*>python_string
adress_in_c = <Py_intptr_t>ptr
address_from_void = adress_in_c        # address_from_void is a python int

ptr2 = <PyObject*>python_string
address_in_c2 = <Py_intptr_t>ptr2
address_from_PyObject = address_in_c2  # address_from_PyObject is a python int

assert address_from_void == address_from_PyObject == id(python_string)

print(<object>ptr)                     # Prints "foo"
print(<object>ptr2)                    # prints "foo"
```
This code is failing to compile with following error:

```
ptr2 = <PyObject*>python_string
      ^
------------------------------------------------------------

test.pyx:12:7: Cannot convert 'PyObject *' to Python object
```

Maybe I understood it incorrectly.

Reply:
Type inference only applies to functions, not the module global code (which only has Python objects in a dict, except for explicitly declared C variables).

Reply:
Hmm so I don't understand. I am still learning cython so I don't understand everything, but it means that I should move cython language version to `main()` function in order to make variables local?